crates/rpc/rpc-convert/src/transaction.rs

```diff
@@ -382,6 +382,66 @@ impl TryIntoTxEnv<TxEnv> for TransactionRequest {
     }
 }
 
+/// Converts rpc transaction requests into transaction environment.
+pub trait TxEnvConverter<RpcReq, T>: Debug + Send + Sync + Unpin + Clone + 'static {
+    /// An associated error that can occur during conversion.
+    type Error;
+
+    /// Converts a rpc transaction request into a transaction environment.
+    fn convert_tx_env<Spec>(
+        &self,
+        request: RpcReq,
+        cfg_env: &CfgEnv<Spec>,
+        block_env: &BlockEnv,
+    ) -> Result<T, Self::Error>;
+}
+
+impl<RpcReq, T> TxEnvConverter<RpcReq, T> for ()
+where
+    RpcReq: TryIntoTxEnv<T>,
+{
+    type Error = RpcReq::Err;
+
+    fn convert_tx_env<Spec>(
+        &self,
+        request: RpcReq,
+        cfg_env: &CfgEnv<Spec>,
+        block_env: &BlockEnv,
+    ) -> Result<T, Self::Error> {
+        request.try_into_tx_env(cfg_env, block_env)
+    }
+}
+
+/// Converts rpc transaction requests into transaction environment using a closure.
+/// Uses unsafe transmute to cast `&CfgEnv<Spec>` to `&CfgEnv<()>`, allowing closures
+/// to access all `CfgEnv` fields.
+impl<F, RpcReq, T, E> TxEnvConverter<RpcReq, T> for F
+where
+    F: Fn(RpcReq, &CfgEnv<()>, &BlockEnv) -> Result<T, E>
+        + Debug
+        + Send
+        + Sync
+        + Unpin
+        + Clone
+        + 'static,
+    RpcReq: Clone,
+    E: std::error::Error + Send + Sync + 'static,
+{
+    type Error = E;
+
+    fn convert_tx_env<Spec>(
+        &self,
+        request: RpcReq,
+        cfg_env: &CfgEnv<Spec>,
+        block_env: &BlockEnv,
+    ) -> Result<T, Self::Error> {
+        //TODO: check if we should use only the chain id - and some other specific fields
+        // or the whole `CfgEnv`
+        let cfg_env_base = unsafe { std::mem::transmute::<&CfgEnv<Spec>, &CfgEnv<()>>(cfg_env) };
```

Comment:
don't think we can do this because there's no guarantee that spec is sized

do we need this closure impl?, or can we add another generic instead of () for the spec